### PR TITLE
Check if ready to deploy 

### PR DESCRIPTION
**Pull request by eeren**

### CODE REVIEW

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs [hunk 1]
```diff
@@ -1,81 +1,100 @@
 ﻿using System;
 using System.Collections.Generic;
+using System.Web.Services.Protocols;
 using System.Text.RegularExpressions;
 using System.Xml;
 
 namespace CurrencyExchangeServiceWCF
 {
-
-    public class Service : IService
```
Comment (JamHaven) [inline CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs:9]: Why did this change fom Service, to Service1?

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs [hunk 2]
```diff
@@ -1,81 +1,100 @@
...
-        public string ConvertCurrency(string toCurrency, string value)
+        SoapException soapException ;
+        public Response<string> ConvertCurrency(ConvertCurrenyRequest req)
         {
-            string currencyInputPattern = @"^[a-zA-Z]{3}$";
-            string valueInputPattern = @"^[,.0-9]*$";
+            try
+            {
+                Response<string> r = new Response<string>();
+                if (!req.autHeader.password.Equals("pa$$w0rd") || !req.autHeader.username.Equals("Admin")){
```
Comment (JamHaven) [inline CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs:20]: I know you want to change this later, but could you create an issue so it won't be forgotten?

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs [hunk 3]
```diff
@@ -1,81 +1,100 @@
...
 
 
-            bool isCurrencyValid = Regex.IsMatch(toCurrency, currencyInputPattern);
-            bool isValueValid = Regex.IsMatch(value, valueInputPattern);
+                if (String.IsNullOrEmpty(value) || String.IsNullOrEmpty(toCurrency))
+                {
+                    r.ReturnValue = "Input cannot be NULL!";
+                    return r;
+                }
+                toCurrency = toCurrency.ToUpper();
```
Comment (JamHaven) [inline CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs:38]: Redundant to line 26

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs [hunk 4]
```diff
@@ -1,81 +1,100 @@
...
     {
-        public string ConvertCurrency(string toCurrency, string value)
+        SoapException soapException ;
+        public Response<string> ConvertCurrency(ConvertCurrenyRequest req)
         {
-            string currencyInputPattern = @"^[a-zA-Z]{3}$";
-            string valueInputPattern = @"^[,.0-9]*$";
+            try
+            {
+                Response<string> r = new Response<string>();
```
Comment (JamHaven) [inline CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs]: Please stick to the naming conventions. Pick meaningful variable names.

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs [hunk 5]
```diff
@@ -1,81 +1,100 @@
...
-                return "Currency not available in the CurrencyConverter currencies list.";
+                if (!isCurrencyValid || !isValueValid)
+                {
+                    r.ReturnValue = "Input cannot be NULL!";
+                    return r;
+                }
 
-            }
+                double ConversionRateDouble;
+                string ConversionRateString;
```
Comment (JamHaven) [inline CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs]: Declarations are all over the place. Please refactor this with more structure.

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs [hunk 6]
```diff
@@ -1,81 +1,100 @@
...
-            string valueInputPattern = @"^[,.0-9]*$";
+            try
+            {
+                Response<string> r = new Response<string>();
+                if (!req.autHeader.password.Equals("pa$$w0rd") || !req.autHeader.username.Equals("Admin")){
+                    r.ReturnValue = "authentication not valid"; 
+                    return r;
+                }
+                String toCurrency = req.toCurrency;
+                String value = req.ccValue;
```
Comment (JamHaven) [inline CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs:25]: Please adapt the variable name to a meaningful name.

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs [hunk 7]
```diff
@@ -1,81 +1,100 @@
...
+                double ConversionRateDouble;
+                string ConversionRateString;
+            
+                if (toCurrency.Equals("USD"))
+                {
+                    r.ReturnValue = value;
+                    return r;
+                }
+                if (toCurrency.Equals("EUR"))
+                    toCurrency = "USD";
```
Comment (JamHaven) [inline CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs:56]: Is this correct? This is a hardcoded change of the incoming parameter

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs [hunk 8]
```diff
@@ -1,81 +1,100 @@
...
+                soapException = new SoapException("Wert konnte nicht umgerechnet werden.", SoapException.ClientFaultCode, "");
+                throw soapException;
             }
-
-            return value;
-
         }
 
-        private string GetActualConversionRate(string currency)
+        private string GetActualConversionRate(string Currency)
```
Comment (JamHaven) [inline CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs:91]: This changes is against the naming convention. Parameters should start with a lower case letter.

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs [hunk 9]
```diff
@@ -85,50 +104,70 @@ private string GetActualConversionRate(string currency)
 
                         while (reader.MoveToNextAttribute())
                         {// Read the attributes.
-                            if (reader.Value.Equals(currency))
+                            if (reader.Value.Equals(Currency))
```
Comment (JamHaven) [inline CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs:104]: This changes is against the naming convention. Parameters should start with a lower case letter.

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs [hunk 10]
```diff
@@ -85,50 +104,70 @@ private string GetActualConversionRate(string currency)
...
             }
             rate = rate.Replace(".", ",");
             return rate;
         }
 
-        public List<string> GetCurrencyCodes()
+        public Response<List<string>> GetCurrencyCodes(GetCurrencyCodesRequest req)
         {
-            List<string> currencyList = new List<string>();
+            Response<List<string>> r = new Response<List<string>>();
```
Comment (JamHaven) [inline CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs]: Please use the correct naming conventions

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs [hunk 11]
```diff
@@ -85,50 +104,70 @@ private string GetActualConversionRate(string currency)
...
+            if (!req.autHeader.password.Equals("pa$$w0rd") || !req.autHeader.username.Equals("Admin"))
```
Comment (JamHaven) [inline CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs:125]: I know you want to change this later, but could you create an issue so it won't be forgotten?

**Review by JamHaven — CHANGES_REQUESTED** [review]
I should have requested these changes before, but the code is a mess and all over the place.
Naming conventions are not upheld and a few functions to structure the code and make it more readable is required. Also, an issue for the password changes, so it won't be forgotten, should be created before the merge is conducted.

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs (hunk 2 above)
Comment (JamHaven) [inline CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs:20]: Also, doseen't this mean the password will be sent without base64 encoding? Or is this done automatically by the framework?

Comment (JamHaven) [reply CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs:20]: Doesn't this mean the password is not sent  via base64 encoding? or is this done by the framework automatically?

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs [hunk 12]
```diff
@@ -1,81 +1,100 @@
...
-                toCurrency = "USD";
+                bool isCurrencyValid = Regex.IsMatch(toCurrency, currencyInputPattern);
+                bool isValueValid = Regex.IsMatch(value, valueInputPattern);
 
-            if (!GetCurrencyCodes().Contains(toCurrency))
-            {
-                return "Currency not available in the CurrencyConverter currencies list.";
+                if (!isCurrencyValid || !isValueValid)
+                {
+                    r.ReturnValue = "Input cannot be NULL!";
```
Comment (JamHaven) [inline CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs]: Wrong error message. In this case the input is wrong, not NULL

**Review by JamHaven — CHANGES_REQUESTED** [review]
Forgott some stuff

**Comment by JamHaven** [comment]
I also forgot that in the Service.svc Service is still annotated as Service, but this changed in the code to Service1

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/IService.cs [hunk 13]
```diff
@@ -1,37 +1,74 @@
...
+     * Date model for the authentication header 
+     */
     [DataContract]
-    public class CompositeType
+    public class AuthHeader
     {
-        [DataMember]
-        public bool BoolValue { get; set; } = true;
+        public string username;
+        public string password;
```
Comment (JamHaven) [inline CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/IService.cs:57]: should be private

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/IService.cs [hunk 14]
```diff
@@ -1,37 +1,74 @@
...
+    }
+    [MessageContract]
+    public class GetCurrencyCodesRequest 
+    {
+        [MessageBodyMember]
+        public AuthHeader autHeader { get; set; }
+    }
+    
+    /**
+     * The return value can use for e
```
Comment (JamHaven) [inline CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/IService.cs:41]: The comment is broken

**Review by JamHaven — CHANGES_REQUESTED** [review]
Service model needs to improved a little

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs (hunk 2 above)
Comment (eeren) [reply CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs:20]: Ok but this is fine for our purpose!

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs (hunk 1 above)
Comment (eeren) [reply CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs:9]: i set the service name again of Service

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs (hunk 3 above)
Comment (eeren) [reply CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs:38]: fixed

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs (hunk 4 above)
Comment (eeren) [reply CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs]: fixed

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs (hunk 3 above)
Comment (erdoganeren) [reply CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs:38]: fixed

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs (hunk 5 above)
Comment (eeren) [reply CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs]: fixed

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs (hunk 7 above)
Comment (eeren) [reply CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs:56]: this is a algotithmus from Christian, i gues he will map the EUR to USD or EUR = USD

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/IService.cs (hunk 13 above)
Comment (eeren) [reply CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/IService.cs:57]: public Variables are required in the [DataContract] context

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/IService.cs (hunk 14 above)
Comment (eeren) [reply CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/IService.cs:41]: fixed

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs (hunk 12 above)
Comment (eeren) [reply CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs]: fixed

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs (hunk 8 above)
Comment (eeren) [reply CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs:91]: fixed

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs (hunk 9 above)
Comment (eeren) [reply CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs:104]: fixed

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs (hunk 10 above)
Comment (eeren) [reply CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs]: fixed

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs (hunk 11 above)
Comment (eeren) [reply CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs:125]: fixed

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs (hunk 2 above)
Comment (eeren) [reply CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs:20]: see below

File: CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs (hunk 7 above)
Comment (JamHaven) [reply CurrencyExchangeServiceWCF/CurrencyExchangeServiceWCF/Service.svc.cs:56]: I guess this is leagcy code then. but leave it for now

**Review by JamHaven — APPROVED** [review]
All in all fine now.

**Pull request merged**